### PR TITLE
feat: SSO with Auth0 for B2C

### PR DESCRIPTION
**Pull request by dhhyi**
## PR Type

[x] Feature
[x] Refactoring (no functional changes, no API changes)

## What Is the New Behavior?

- SSO support via Auth0

## Does this PR Introduce a Breaking Change?

[x] Yes
[ ] No

## Limitations
- support only B2C users
- No changes of email possible, email provided from IDP is used
- No customer approval supported
- Only one IDP supported
- Redirect to loading page to determine next page
- No creation of customers via Backoffice supported
- When using Auth0 username/password login, the name of the user must be changed from email to something useful, otherwise the user cannot set preferred payments or addresses.

## Open Questions / Issues:
- [x] Do we need to have capability for anonymous shopping (cart + checkout) for Auth0? This would result in a mixed scenario. - Yes
- [x] Auth0 Registration: subheading "Contact Information" differs from Registration Pages subheading "Address". Do we really want to introduce the new localization key?
- [x] Checking if a user is new is done by checking if the ICM generated users generated address hat no first address line. Why generate user in the first place? -> https://jira.intershop.de/browse/IS-31067
- [x] logging in at ICM has to be retried multiple times till it finally succeeds -> https://jira.intershop.de/browse/IS-31115
- [x] ICM always overwrites some of users fields on login (firstName, lastName, ...)

## TODO:
- [x] introduce IdentityProvider abstraction to switch between standard handling and Auth0 handling
- ICM Identity Provider
- [x] abstract APIToken Service in restore.effects
- [x] move all user-related standard handling for API Token to ICM Identity Provider (so it can be "switched off")
- Auth0 Identity Provider
- [x] Login/logout via Auth0
- [x] send retrieved ID Token to ICM to recover/create User
- [x] New users are directed to registration form to complete their profile
- [x] Returning users are directed to the ~~previously visited page~~ myaccount (standard behavior)
- [x] abstract TAC component
- [x] Add to cart redirects to login
- [x] Disable edit email in profile settings
- [x] Hide edit password in profile settings
- [x] missing/invalid fields error for registration (maybe default preferredLanguage missing?)
- [x] register link should also lead to Auth0
- [x] mixed scenario for Auth0 identity provider (anonymous shopping, merge basket)
- [x] remove workaround for IS-30270 (CMS w/o pgid)
- [x] Documentation
- [x] switch identity provider via Multi-Channel

### CODE REVIEW

**Review by MaxKless — COMMENTED** [review]
Code looks good, will do more testing next

File: docs/concepts/sso.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    }
+  },
+```
+
+For production this configuration should be provided via environment variables to the SSR process (see [Building and Running Server-Side Rendering][ssr-startup]).
+The usage of identity providers can also be provided in the multi-channel configuration (see [Building and Running nginx Docker Image][nginx-startup]).
+
+# Further References
+
+- PWA
```
[the remark this answers is not in the document]
Comment (dhhyi) [reply docs/concepts/sso.md:33]: 👍

File: docs/guides/sso-auth0.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+```typescript
+  identityProvider: 'MyProvider',
+  identityProviders: {
+    'MyProvider': {
+      type: 'auth0',
+      domain: 'some-domain.auth0.com',
+      clientID: 'ASDF12345',
+    }
+  },
+```
```
[the remark this answers is not in the document]
Comment (dhhyi) [reply docs/guides/sso-auth0.md:25]: 👍

**Review by MaxKless — CHANGES_REQUESTED** [review]
image
I ran into this issue when using multiple channels that share a user (see multi channel configuration at the bottom). 
**Steps to reproduce:**
- go to pwa.net/de
- sign in using Auth0
- navigate to pwa.net/us
- log out 
-> see the error above. The logout still works and when you go back to the pwa, you are logged out. 

**Configuration**

```
# export appropriate ICM_BASE_URL and use with
# docker-compose -f templates/docker-compose.yml up --build

version: '3'
services:
  pwa:
    build:
      context: ..
      args:
        # configuration: local
        configuration: production
        serviceWorker: 'false'
    environment:
      - LOGGING=on
      - ICM_BASE_URL
      # - PROXY_ICM=true
      - TRUST_ICM=true
      # - PROMETHEUS=on
      - IDENTITY_PROVIDERS={"Auth0":{"type":"auth0","domain":"dev-9pqwi8uh.eu.auth0.com","clientID":"V2xvLUo3YSfGdCmJz7u381dSevoVkOso"}}
  nginx:
    build: ../nginx
    depends_on:
      - pwa
    environment:
      UPSTREAM_PWA: 'http://pwa:4200'
      # DEBUG: 1
      CACHE: 0
      PAGESPEED: 0
      # PROMETHEUS: 1
      # COMPRESSION: 0
      # DEVICE_DETECTION: 0
      # MULTI_CHANNEL_SOURCE: env:///ASDF?type=application/yaml
      MULTI_CHANNEL: |
        .+\.net:
            - baseHref: /de
              lang: de_DE
              channel: inSPIRED-inTRONICS-Site
              identityProvider: Auth0
            - baseHref: /us
              lang: en_US
              channel: inSPIRED-inTRONICS-Site
              identityProvider: Auth0
        .+\.de:
            channel: inSPIRED-inTRONICS-Site
            lang: de_DE
            identityProvider: Auth0
        .+\.com:
            channel: inSPIRED-inTRONICS_Business-Site
            features: quoting,businessCustomerRegistration,advancedVariationHandling
            theme: "blue|688dc3"
        .+\.fr:
            channel: inSPIRED-inTRONICS-Site
            lang: fr_FR
            application: smb-responsive
            features: quoting
            theme: "blue|688dc3"

    ports:
      - '4200:80'
      # - '9113:9113'

```

**Comment by dhhyi** [comment]
>       MULTI_CHANNEL: |
>         .+\.net:
>             - baseHref: /de
>               lang: de_DE
>               channel: inSPIRED-inTRONICS-Site
>               identityProvider: Auth0
>             - baseHref: /us
>               lang: en_US
>               channel: inSPIRED-inTRONICS-Site
>               identityProvider: Auth0

The error you see comes from the Auth0 configuration, because `/us` routes are not configured to be valid URLs. This has to be done in the Auth0 backoffice.

What would be your expectation at this point?

**Review by marschmidt89 — DISMISSED** [review]
documentation review done

**Review by MaxKless — APPROVED** [review]

**Pull request merged**